ReadME.md
# Simple React Table

A Simple React Component

[DEMO](http://kimbugp.github.io/react_table)

## Install

```sh
npm install --save sortable-react-table
```

## Settings

```js
import React from "react";
import ReactDom from "react-dom";
import SimpleTable from "sortable-react-table";

class App extends React.Component {
  render() {
    return (
      <SimpleTable
        data={[
          { id: 1, name: "simon" },
          { id: 2, name: "peter" }
        ]}
        title="simple table"
        sortables={[{ name: "id", position: "left" }]}
      ></SimpleTable>
    );
  }
}

ReactDom.render(<App />, document.getElementById("App"));
```

## Sample

![Sample](/sample.png)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### 0.0.11 (2020-02-19)

### 0.0.10 (2020-02-19)

### 0.0.9 (2020-02-19)

### [0.0.9](https://github.com/kimbugp/react_table/compare/v0.0.6...v0.0.9) (2020-02-19)


### Bug Fixes

* **update:** update scope ([7f3ffab](https://github.com/kimbugp/react_table/commit/7f3ffabb368dc075f5b52622c7cd2cf0f26ff937))

### [0.0.8](https://github.com/kimbugp/react_table/compare/v0.0.6...v0.0.8) (2020-02-19)


### Bug Fixes

* **update:** update scope ([7f3ffab](https://github.com/kimbugp/react_table/commit/7f3ffabb368dc075f5b52622c7cd2cf0f26ff937))

### [0.0.7](https://github.com/kimbugp/react_table/compare/v0.0.6...v0.0.7) (2020-02-19)

### [0.0.6](https://github.com/kimbugp/react_table/compare/v0.0.5...v0.0.6) (2020-02-19)


### Bug Fixes

* **token:** add github token ([d4f9a79](https://github.com/kimbugp/react_table/commit/d4f9a792125a45949fd2d27f761cb6676a6d6a5e))

### [0.0.5](https://github.com/kimbugp/react_table/compare/v0.0.4...v0.0.5) (2020-02-19)


### Features

* **lint:** add commit lint ([c255dee](https://github.com/kimbugp/react_table/commit/c255dee1bbfbb6172ce92b8368b8b4ee89ca8b1c))

### [0.0.3](https://github.com/kimbugp/react_table/compare/v0.0.4...v0.0.3) (2020-02-19)


### Features

* **lint:** add commit lint ([c255dee](https://github.com/kimbugp/react_table/commit/c255dee1bbfbb6172ce92b8368b8b4ee89ca8b1c))


src/App.test.js
import React from 'react';
import ReactDOM from 'react-dom';
import expect from 'expect';
import App from './App';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { mount, shallow } from 'enzyme';
import renderer from 'react-test-renderer';

configure({ adapter: new Adapter() });

describe('AppComponent', () => {
  it('check if parameterState changes', () => {
    const wrapper = mount(<App />);
    wrapper.instance().sortByParameter("age");
    expect(wrapper.state().parameterState).toEqual("age");
  });
});


src/components/Radio.test.js
import React from "react";
import ReactDOM from "react-dom";
import expect from "expect";
import App from "../App";
import Radio from "./Radio";
import { configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { mount, shallow } from "enzyme";
import renderer from "react-test-renderer";

configure({ adapter: new Adapter() });

describe("RadioComponent", () => {
  it("check if radio state changes by name", () => {
    const app = mount(<App />);
    const radio = app.find(Radio).get(0);
    radio.props.sortBy("name");
    expect(app.state().parameterState).toEqual("name");
  });

  it("check if radio state changes by age", () => {
    const app = mount(<App />);
    const radio = app.find(Radio).get(0);
    radio.props.sortBy("age");
    expect(app.state().parameterState).toEqual("age");
  });
});


src/components/Table.test.js
import React from "react";
import ReactDOM from "react-dom";
import expect from "expect";
import App from "../App";
import Table from "./Table";
import { configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { mount } from "enzyme";

configure({ adapter: new Adapter() });

describe("TableComponent", () => {
  it("check date comparison", () => {
    const wrapper = mount(
      <Table
        sortParameter={"birth"}
        data={[
          {
            name: "Ross Geller",
            birth: "11/30/2010"
          }
        ]}
      />
    );
    let res = wrapper.instance().compareVariables(
      {
        name: "Ross Geller",
        birth: "11/30/2010"
      },
      {
        name: "Mike Ross",
        birth: "09/20/1992"
      }
    );
    expect(res).toEqual(1);
  });

  it("check name comparison", () => {
    const wrapper = mount(
      <Table
        sortParameter={"name"}
        data={[
          {
            name: "Ross Geller",
            birth: "11/30/2010"
          }
        ]}
      />
    );
    let res = wrapper.instance().compareVariables(
      {
        name: "Ross Geller",
        birth: "11/30/2010"
      },
      {
        name: "Mike Ross",
        birth: "09/20/1992"
      }
    );
    expect(res).toEqual(1);
  });

  it("check number of rows rendered", () => {
    const table = mount(
      <Table
        sortParameter={"name"}
        data={[
          {
            name: "Ross Geller",
            birth: "11/30/2010"
          }
        ]}
      />
    );
    let el = table.find("tr");
    expect(el.length).toEqual(2);
  });

  it("check random row for sort by name", () => {
    let data = [
      { name: "Shafaet Asharaf", birth: "03/12/1934" },
      { name: "HariShankaran Karunanidhi", birth: "07/06/1986" },
      { name: "Aakansha Doshi", birth: "01/12/1988" },
      { name: "Vivek Ravishankaran", birth: "01/23/1988" },
      { name: "Vishal Kothari", birth: "12/12/1990" },
      { name: "Hemant Kumar", birth: "12/31/1990" },
      { name: "Ashley Bose", birth: "01/16/1992" },
      { name: "Suman Kumar", birth: "03/12/2001" }
    ];
    const table = mount(<Table sortParameter={"name"} data={data} />);
    table.instance().props.data.sort(table.instance().compareVariables);
    expect(table.instance().sortItems(data)[3]).toEqual({
      name: "Hemant Kumar",
      birth: "12/31/1990"
    });
  });
});


service-worker.js
"use strict";var precacheConfig=[["/react_table/index.html","607b0daf0a5c6b78ac3cf8246051bfad"],["/react_table/static/css/main.af16badc.css","e5d91dc94490d323d84d5126c2d38955"],["/react_table/static/js/main.a0e61170.js","4af5966c8df6f81634ec79b3338ce7cf"]],cacheName="sw-precache-v3-sw-precache-webpack-plugin-"+(self.registration?self.registration.scope:""),ignoreUrlParametersMatching=[/^utm_/],addDirectoryIndex=function(e,t){var n=new URL(e);return"/"===n.pathname.slice(-1)&&(n.pathname+=t),n.toString()},cleanResponse=function(t){return t.redirected?("body"in t?Promise.resolve(t.body):t.blob()).then(function(e){return new Response(e,{headers:t.headers,status:t.status,statusText:t.statusText})}):Promise.resolve(t)},createCacheKey=function(e,t,n,r){var a=new URL(e);return r&&a.pathname.match(r)||(a.search+=(a.search?"&":"")+encodeURIComponent(t)+"="+encodeURIComponent(n)),a.toString()},isPathWhitelisted=function(e,t){if(0===e.length)return!0;var n=new URL(t).pathname;return e.some(function(e){return n.match(e)})},stripIgnoredUrlParameters=function(e,n){var t=new URL(e);return t.hash="",t.search=t.search.slice(1).split("&").map(function(e){return e.split("=")}).filter(function(t){return n.every(function(e){return!e.test(t[0])})}).map(function(e){return e.join("=")}).join("&"),t.toString()},hashParamName="_sw-precache",urlsToCacheKeys=new Map(precacheConfig.map(function(e){var t=e[0],n=e[1],r=new URL(t,self.location),a=createCacheKey(r,hashParamName,n,/\.\w{8}\./);return[r.toString(),a]}));function setOfCachedUrls(e){return e.keys().then(function(e){return e.map(function(e){return e.url})}).then(function(e){return new Set(e)})}self.addEventListener("install",function(e){e.waitUntil(caches.open(cacheName).then(function(r){return setOfCachedUrls(r).then(function(n){return Promise.all(Array.from(urlsToCacheKeys.values()).map(function(t){if(!n.has(t)){var e=new Request(t,{credentials:"same-origin"});return fetch(e).then(function(e){if(!e.ok)throw new Error("Request for "+t+" returned a response with status "+e.status);return cleanResponse(e).then(function(e){return r.put(t,e)})})}}))})}).then(function(){return self.skipWaiting()}))}),self.addEventListener("activate",function(e){var n=new Set(urlsToCacheKeys.values());e.waitUntil(caches.open(cacheName).then(function(t){return t.keys().then(function(e){return Promise.all(e.map(function(e){if(!n.has(e.url))return t.delete(e)}))})}).then(function(){return self.clients.claim()}))}),self.addEventListener("fetch",function(t){if("GET"===t.request.method){var e,n=stripIgnoredUrlParameters(t.request.url,ignoreUrlParametersMatching),r="index.html";(e=urlsToCacheKeys.has(n))||(n=addDirectoryIndex(n,r),e=urlsToCacheKeys.has(n));var a="/react_table/index.html";!e&&"navigate"===t.request.mode&&isPathWhitelisted(["^(?!\\/__).*"],t.request.url)&&(n=new URL(a,self.location).toString(),e=urlsToCacheKeys.has(n)),e&&t.respondWith(caches.open(cacheName).then(function(e){return e.match(urlsToCacheKeys.get(n)).then(function(e){if(e)return e;throw Error("The cached response that was expected is missing.")})}).catch(function(e){return console.warn('Couldn\'t serve response for "%s" from cache: %O',t.request.url,e),fetch(t.request)}))}});

src/components/Radio.js
import React, { Component } from "react";
import { PropTypes } from "prop-types";

class Radio extends Component {
  constructor(props) {
    super(props);
    this.state = {
      radioClick: "name"
    };
  }

  onChange = type => {
    const { id } = type.target;
    this.setState({
      radio: id,
      radioClick: id
    });
    this.props.sortBy(id);
  };
  getButtons = items => {
    return items.map((item, index) => {
      return (
        <Button
          {...item}
          onChange={this.onChange}
          radio={this.state.radio}
          index={index}
          key={index}
        />
      );
    });
  };
  render() {
    return (
      <div className="radioButtons">
        {this.getButtons(this.props.sortables)}
      </div>
    );
  }
}

const Button = props => {
  return (
    <div className={props.position}>
      <input
        type="radio"
        onChange={props.onChange}
        id={props.name}
        checked={props.radio === props.name}
      />
      <label>&nbsp;&nbsp;Sort by {props.name}</label>
    </div>
  );
};

Radio.propTypes = {
  sortBy: PropTypes.func
};

export default Radio;


src/components/Table.js
import React, { Component } from "react";
import { PropTypes } from "prop-types";

class Table extends Component {
  getKeys = () => {
    return Object.keys(this.props.data[0]);
  };

  getHeader = () => {
    let keys = this.getKeys();
    return keys.map((key, index) => {
      return <th key={key}>{key.toUpperCase()}</th>;
    });
  };
  sortItems = items => {
    return items.sort(this.compareVariables);
  };
  getRowsData = () => {
    let items = this.props.data;
    let keys = this.getKeys();
    return this.sortItems(items).map((row, index) => {
      return (
        <tr key={index}>
          <Rows key={index} data={row} keys={keys} />
        </tr>
      );
    });
  };
  isDate = date => {
    return new Date(date) !== "Invalid Date" && !isNaN(new Date(date))
      ? true
      : false;
  };
  getComparator = (data1, data2, param) => {
    let var1 = data1[param];
    let var2 = data2[param];
    if (this.isDate(var1) & this.isDate(var2)) {
      return { var1: new Date(var1), var2: new Date(var2) };
    }
    return { var1, var2 };
  };

  compareVariables = (data1, data2) => {
    let param = this.props.sortParameter;
    let { var1, var2 } = this.getComparator(data1, data2, param);
    if (var1 <= var2) {
      return -1;
    }
    if (var1 >= var2) {
      return 1;
    }
    return 0;
  };

  render() {
    return (
      <div className="table-div">
        <table className="table table-striped table-bordered table-hover full-width">
          <thead>
            <tr>{this.getHeader()}</tr>
          </thead>
          <tbody>{this.getRowsData()}</tbody>
        </table>
      </div>
    );
  }
}

const Rows = props => {
  return props.keys.map((key, index) => {
    return (
      <td key={props.data[key]} id={index}>
        {props.data[key]}
      </td>
    );
  });
};
Table.defaultProps = {
  data: []
};

Table.propTypes = {
  sortParameter: PropTypes.string,
  data: PropTypes.array
};

export default Table;


src/App.js
import React, { Component } from "react";
import Radio from "./components/Radio";
import Table from "./components/Table";
import "./index.css";

class SimpleTable extends Component {
  constructor(props) {
    super(props);
    this.state = {
      parameterState: "name"
    };
  }

  sortByParameter = parameter => {
    this.setState({
      parameterState: parameter
    });
  };

  render() {
    return (
      <div className="container-fluid">
        <center>
          <h1>{this.props.title}</h1>
        </center>
        <Radio sortBy={this.sortByParameter} sortables={this.props.sortables} />
        <Table
          sortParameter={this.state.parameterState}
          data={this.props.data}
        />
      </div>
    );
  }
}

SimpleTable.defaultProps = {
  title: "Simple table",
  sortables: [
    { name: "name", position: "left" },
    { name: "birth", position: "right" }
  ],
  data: [
    {
      name: "John Sina",
      birth: "11/30/2011"
    },
    {
      name: "Barcy Washington",
      birth: "09/16/1992"
    },
    {
      name: "Peter Parker",
      birth: "01/16/1992"
    },
    {
      name: "Jimmy Shergil",
      birth: "12/12/2001"
    },
    {
      name: "Alexander Alfred",
      birth: "02/09/1891"
    },
    {
      name: "Krishna Gupta",
      birth: "12/01/1982"
    },
    {
      name: "Sania Mirza",
      birth: "11/30/2011"
    },
    {
      name: "Lata Pathak",
      birth: "10/31/1999"
    }
  ]
};

export default SimpleTable;
